Crash under maskedURLForBindings

Constellation opened the pull request:
#### 7626eadf0a70b425cf472ec22d17bcb66e16468a
<pre>
Crash under maskedURLForBindings
<a href="https://bugs.webkit.org/show_bug.cgi?id=243086">https://bugs.webkit.org/show_bug.cgi?id=243086</a>
rdar://97409683

Reviewed by Mark Lam, Chris Dumez and Timothy Hatcher.

JSC has a feature materializing Error stack from GC heap thread because of massive memory saving: if we keep
call stack information via JSCell, it can keeps a lot of JSCells from that, and keeping massive amount of memory forever
until an error object is GC-ed.

However, 252418@main introduced a hook into this stack materialization, and this hook function is not designed to be called
from GC heap thread: it is assuming the main thread, and it is generating AtomString. This does not work since GC heap thread
is not a main thread. JSC is intentionally clears AtomStringTable in GC heap threads to catch these errors, and thus we crash now.

This patch fixes Document::maskedURLForBindings so that it creates a URL in a thread-safe way: creating it via LazyNeverDestroyed + std::call_once.
And it is intentionally using StaticStringImpl since which cannot be converted to AtomString and this string can be used from any threads
at the same time. On the other hand, Element::resolveURLStringIfNeeded is called only from the main thread. Thus we make the used string
AtomString in that function so that we can avoid atomizing the same string repeatedly.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::maskedURLStringForBindings):
(WebCore::Document::maskedURLForBindings):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::resolveURLStringIfNeeded const):

Canonical link: <a href="https://commits.webkit.org/252724@main">https://commits.webkit.org/252724@main</a>
</pre>

MenloDorian reviewed (approved):
r=me too

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5627,15 +5627,21 @@ bool Document::shouldMaskURLForBindingsInternal(const URL& urlToMask) const
+    static LazyNeverDestroyed<URL> url;
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [&] {
+        url.construct(maskedURLStringForBindings());
```
Can we ASSERT() that `url.string()` is indeed still backed by a StaticStringImpl? URL wouldn't be thread-safe otherwise. This is a very fragile design given that the URLParser may or may not use the input String as URL.m_string.

Constellation replied:
Sounds good, done!

Constellation committed:
Crash under maskedURLForBindings
https://bugs.webkit.org/show_bug.cgi?id=243086
rdar://97409683

Reviewed by Mark Lam, Chris Dumez and Timothy Hatcher.

JSC has a feature materializing Error stack from GC heap thread because of massive memory saving: if we keep
call stack information via JSCell, it can keeps a lot of JSCells from that, and keeping massive amount of memory forever
until an error object is GC-ed.

However, 252418@main introduced a hook into this stack materialization, and this hook function is not designed to be called
from GC heap thread: it is assuming the main thread, and it is generating AtomString. This does not work since GC heap thread
is not a main thread. JSC is intentionally clears AtomStringTable in GC heap threads to catch these errors, and thus we crash now.

This patch fixes Document::maskedURLForBindings so that it creates a URL in a thread-safe way: creating it via LazyNeverDestroyed + std::call_once.
And it is intentionally using StaticStringImpl since which cannot be converted to AtomString and this string can be used from any threads
at the same time. On the other hand, Element::resolveURLStringIfNeeded is called only from the main thread. Thus we make the used string
AtomString in that function so that we can avoid atomizing the same string repeatedly.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::maskedURLStringForBindings):
(WebCore::Document::maskedURLForBindings):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::resolveURLStringIfNeeded const):

Canonical link: https://commits.webkit.org/252724@main
